File: src/api/resources.ts
import { type SessionHandle, TabLocation, type Resource } from "~/models";
import { homeworkFromWeek, homeworkFromIntervals } from "./private/homework";
import { decodeResource } from "~/decoders/resource";

const decoder = (session: SessionHandle, data: any): Array<Resource> => {
  return data.ListeCahierDeTextes.V.map((resource: any) => decodeResource(resource, session));
};

export const resourcesFromWeek = async (session: SessionHandle, weekNumber: number, extendsToWeekNumber?: number): Promise<Resource[]> => {
  const reply = await homeworkFromWeek(session, TabLocation.Resources, weekNumber, extendsToWeekNumber);
  return decoder(session, reply);
};

export const resourcesFromIntervals = async (session: SessionHandle, startDate: Date, endDate: Date): Promise<Resource[]> => {
  const reply = await homeworkFromIntervals(session, TabLocation.Resources, startDate, endDate);
  return decoder(session, reply).filter((lesson) => startDate <= lesson.endDate && lesson.endDate <= endDate);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/resources.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.private.homeworkFromIntervals
import ink.literate.pawnote.api.private.homeworkFromWeek
import ink.literate.pawnote.decoders.decodeResource
import ink.literate.pawnote.models.Resource
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

fun decoder (session: SessionHandle, data: JsonObject) = data["ListeCahierDeTextes"]!!.jsonObject["V"]!!.jsonArray.map { decodeResource(it.jsonObject, session) }

suspend fun resourcesFromWeek (session: SessionHandle, weekNumber: Int, extendsToWeekNumber: Int? = null): List<Resource> {
    val reply = homeworkFromWeek(session.information, TabLocation.Resources, weekNumber, extendsToWeekNumber)
    return decoder(session, reply)
}

suspend fun resourcesFromIntervals (session: SessionHandle, startDate: LocalDateTime, endDate: LocalDateTime): List<Resource> {
    val reply = homeworkFromIntervals(session, TabLocation.Resources, startDate, endDate)
    return decoder(session, reply).filter { lesson -> lesson.endDate in startDate..endDate }
}